numpy/core/src/multiarray/ctors.c

```diff
@@ -2351,7 +2351,11 @@ PyArray_FromStructInterface(PyObject *input)
 
     attr = PyArray_LookupSpecial_OnInstance(input, "__array_struct__");
     if (attr == NULL) {
-        return Py_NotImplemented;
+        if (PyErr_Occurred()) {
+            return NULL;
+        } else {
+            return Py_NotImplemented;
+        }
```

Comment:
Why don't you need to now change all the places `PyArray_LookupSpecial_OnInstance` and `PyArray_LookupSpecial_OnInstance` are called to check for errors?

Reply:
I don't know. This change was enough to make my tests pass, so I stopped with this. If you would like me to also change this for every other code path, I would need examples for triggering these paths to verify that my changes actually do something.

Reply:
Unfortunately we do now need to change every call site to either handle it or call `PyErr_Clear()`, else you risk a `SystemError` occurring.

Reply:
In the interest of speed, I suggest you:
* Create a new tracking issue, "propagate errors where appropriate from PyArray_LookupSpecial"
* Find every occurence of `PyArray_LookupSpecial` and `PyArray_LookupSpecial_OnInstance`, and in the `if (ret == NULL)` branch, add `PyErr_Clear();  // TODO[gh-XYZ]: propagate this?

Reply:
done, see #14801.

Could you look over the changes in `binop_override.h` and `arrayfunction_override.c`? I'm not sure about those, especially the change from `if (attr)` to `if (attr != NULL)`.

Reply:
Those changes are fine, thanks for checking. I think we could make the comments a little less cryptic, see my new comment thread below, marking this one as resolved.